_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,145 @@
+
+To continue that point, I often seem teams that will have a "development" environment in their pipeline. This is a step above developers running entire copies of the system in their local environment, but it still comes with issues. It can be difficult to accurately determine exactly what is running in this environment when multiple developers are deploying changes to such an environment, which can make it difficult to ensure working code is making it's way towards production. This is often ensures that a manual gate is introduced and that it never leaves. If developers must be able to run an isolated copy of the system, it is better to implement this via developer sandboxes or dynamic environments that are free of the pipeline. Recognise that the purpose of such an environment isn't to safely deploy code to production, but to aide the developer in feature development, and as such it doesn't belong in the production pipeline.
+
+# A Flexible Model For Software Delivery Pipelines
+
+Assume we are deploying software-as-a-service. This could be an internal or customer facing service.
+
+1. Customers only consume the service from code that has been merged to the master branch.
+
+2. Features that are ready for consumption are merged to the master branch.
+
+3. We have one pipeline that is 'subscribed' to the master branch.
+
+4. All environments in the pipeline exist for the sole purpose of ensuring that changes do not degrade the customer-facing instance.
+
+5. Developers may create a simplified version of the master pipeline - except there is no final "production" stage, and there is only one pre-production stage that executes all tests against that one environment. There is no automated rollback - developers are expected to fix-forward.
+
+6. UAT, training, and copies of the service exist on their own branches with their own copies of the pipeline subscribed to them. Changes may be merged into these branches before or after they reach master depending on the circumstances.
+
+Illustrated, we might expect this to look something like the following graphic.
+
+<center><img src="/img/manual-gates/model.png" /></center>
+<br/>
+
+This has quite a few things going for it.
+
+- We have decoupled environments that have potential manual verification (e.g. UAT) from the path to production.
+- There is a clear method for creating additional environments (e.g. UAT, dev) such that they don't impact the path to production.
+- It can support different branching models with little impact to any current delivery pipelines that may be in use.
+- Decide you don't need additional environments besides pre-prod and prod (e.g. true Trunk-Based Development)? Fine. Just don't create additional environments subscribed to branches.
+
+# Conclusion
```

Comment:
The conclusion is good - but to re-iterate my point above, `# A Flexible Model For Software Delivery Pipelines`doesn't really fit in with it.